File: homework/quang_tu/path-sum-iii.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int pathSum(TreeNode* root, int targetSum) {
        stack<TreeNode*> st;
        int result = 0;

        while(root || st.size()) {
            while (root) {
                st.push(root);
                result += dfs(root, targetSum);
                root = root->left;
            }

            root = st.top();
            st.pop();

            root = root->right;
        }

        return result;
    }

    long dfs(TreeNode* root, long target) {
        if (!root)
            return 0;

        return (root->val == target) + dfs(root->left, target - root->val) +
            dfs(root->right, target - root->val);
    }
};

File: homework/Van_Le/homework7.1/PathSumIII.cs
public class PathSumIII
{
    public static int retVal;
    public static Dictionary<long, int> dict;
    public int PathSum(TreeNode root, int targetSum) {
        retVal = 0;
        dict = new Dictionary<long, int>();
        dict[0] = 1;
        dfs(root, targetSum, 0);
        return retVal;
    }
    
    public void dfs(TreeNode root, int target, long currentSum){
        if(root == null)return;
        
        currentSum += root.val;
        if(dict.ContainsKey(currentSum - target)) {
            retVal += dict[currentSum - target];
        }
        
        if(!dict.ContainsKey(currentSum))
            dict[currentSum] = 0;
        dict[currentSum]++;
            
        dfs(root.left, target, currentSum);
        
        dfs(root.right, target, currentSum);
        
        dict[currentSum]--;
    }
}